### ISSUE TITLE
Konva 5 Transformer

### ISSUE THREAD

**Issue by dhalldmg** [konvajs/konva#900]
It appears there is a bug in the way absolutePosition is calculated in the new transformer.  It doesn't seem to take into account the layer offset from the stage.
In the 2 images you can see the differences in the console.  The only difference is the Konva version used.

![4](https://user-images.githubusercontent.com/58263614/80646258-65455400-8a3a-11ea-819d-93766231cf33.JPG)
![5](https://user-images.githubusercontent.com/58263614/80646259-65455400-8a3a-11ea-9fb9-b2d93a7aca3c.JPG)

**Comment by lavrton** [comment]
Can you make a demo?

**Comment by lavrton** [comment]
I tried this: https://jsbin.com/xovucudizo/edit?html,js,console,output

**Comment by dhalldmg** [comment]
I thought it may be related to my stage being draggable and scale not 1:1, but as I'm trying to create a demo those factors are not affecting it.   If I'm able to figure out what I'm currently missing to recreate it I'll share the demo.

**Comment by dhalldmg** [comment]
I think this helps show the problem.  
Drag the red rectangle a little to expose the gray background.  Then drag the green square and compare absolutePosition vs position.

Then change to konva5 and repeat the above steps. 

https://jsbin.com/woyoyop/edit?html,js,console,output

**Comment by lavrton** [comment]
I see it. https://jsbin.com/weciyelose/1/edit?html,js,console,output

So as I can see `absolutePosition()` returns correct result.
In `v5` `position()` returns the same value as `absolutePosition()`. That may be unexpected, but I am not sure if we will fix that issue. From `v5` `Konva.Transformer` works in absolute space.

**Comment by dhalldmg** [comment]
Thanks for looking into it.
I'm curious if you have a better suggestion for this goal then -   I need to store a group's position and dimensions.   By setting up a transformer and attaching it to the group I could use its position and dimensions perfectly.  Am I missing a feature or property in Konva that would give me the same result?

**Comment by lavrton** [comment]
Group has its own `x`, `y`, `scaleX`, `scaleY`. Why not use them?

**Comment by dhalldmg** [comment]
I am mainly using a transformer to get width/height of a box containing all children of the group.    I used the transformer X Y then at the same time and noticed the position difference first.   Testing out the transformer width/height it seems impacted by the stage scaling.   https://jsbin.com/tukizuyeco/edit?html,js,console,output

**Comment by lavrton** [comment]
I am still not sure what do you want.

Can relative position/size of transforming shapes?

**Comment by dhalldmg** [comment]
I'm sorry for the confusion.
My end goal is to use the relative position/size of a group as the dimensions for drawing out a dynamic imagemap.   The Konva node is a property of my object that contains a user defined link.  
In v4 I was able to get these dimensions by creating a transformer and attaching the group to it.

If transformers are going to remain in absolute space then maybe an enhancement of group could be the width/height properties being used as the dimensions that contain all child nodes of the group?

**Comment by lavrton** [comment]
Sorry, I still don't get it. 😐

So are you used transformer to just get width/height?

**Comment by dhalldmg** [comment]
https://jsbin.com/duhicov/edit?html,js,console,output

I have this basic setup where the white BG is my users design space.
My object contains a 'link' property that a user can put a URL into, and again, the Konva node is a property of my object.

When they save the canvas as a PNG I need to to capture the position and size of the Konva node in my object that has the link so that I can create a HTML imagemap.  I do this by creating a transformer during the save process and store it's position/size in my object.

In this example consider the group as the object I need to make a link to konvajs.org for.
The transformer's absolute position, width and height are the exact dimensions I need and use for this output

<map name="imgmap">
<area shape="rect" coords="50,50,225,250" href="https://konvajs.org/" alt="Konva">
</map>

<img src="(canvas data)" usemap="#imgmap">



In version 4 the stage scale and zoom had no impact on those transformer dimensions.   In version 5 since it is in a different space the results are not right for my goal.   I understand the change in 5, but I don't know the best way to get back the dimensions I need.

**Comment by lavrton** [comment]
So probably you can just calculate size manually, by applying scale to `width` and `height`:

```js
tr.width() / stage.scaleX();
tr.height() / stage.scaleY();
```

Or even better, just use relative `getClientRect()` (so stage scale is ignored):

```js
var box = group.getClientRect({ relativeTo: stage  });
```

**Comment by dhalldmg** [comment]
I was literally just syncing the change you suggested and wanted to let you know that it worked.I wasn’t aware that getClientRect could accept a relativeTo option.  Thank you for your input and help on that From: Anton LavrenovSent: Friday, May 15, 2020 1:33 PMTo: konvajs/konvaCc: dhalldmg; AuthorSubject: Re: [konvajs/konva] Konva 5 Transformer (#900) Closed #900.—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.
